soroban-sdk-macros/src/lib.rs

```diff
@@ -95,6 +97,67 @@ pub fn contractimpl(_metadata: TokenStream, input: TokenStream) -> TokenStream {
     }
 }
 
+#[derive(Debug, FromMeta)]
+struct MetadataArgs {
+    key: String,
+    val: String,
+}
+
+#[proc_macro]
+pub fn contractmeta(metadata: TokenStream) -> TokenStream {
+    let args = parse_macro_input!(metadata as AttributeArgs);
+    let args = match MetadataArgs::from_list(&args) {
+        Ok(v) => v,
+        Err(e) => return e.write_errors().into(),
+    };
+
+    let gen = {
+        let key: StringM<10> = match args.key.clone().try_into() {
+            Ok(k) => k,
+            Err(e) => {
+                return Error::new(Span::call_site(), e.to_string())
+                    .into_compile_error()
+                    .into()
+            }
+        };
+
+        let val: StringM<256> = match args.val.try_into() {
+            Ok(k) => k,
+            Err(e) => {
+                return Error::new(Span::call_site(), e.to_string())
+                    .into_compile_error()
+                    .into()
+            }
+        };
+
+        let meta_v0 = ScMetaV0 { key, val };
+        let meta_entry = ScMetaEntry::ScMetaV0(meta_v0);
+        let metadata_xdr: Vec<u8> = match meta_entry.to_xdr() {
+            Ok(v) => v,
+            Err(e) => {
+                return Error::new(Span::call_site(), e.to_string())
+                    .into_compile_error()
+                    .into()
+            }
+        };
+
+        let metadata_xdr_lit = proc_macro2::Literal::byte_string(metadata_xdr.as_slice());
+        let metadata_xdr_len = metadata_xdr.len();
+
+        let ident = format_ident!("{}", args.key);
```

Comment:
I think we should prefix the identifier for the variable so it is less likely to collide with user define types, functions, etc. For example:
```suggestion
        let ident = format_ident!("__CONTRACT_META_{}", args.key);
```

Reply:
It will be possible for keys to exist that contain characters that aren't valid rust identifiers. If it was trivial to hash the key, that'd be another way, but that'd required pulling in another dependency for this crate. There might be another way, such as converting the key into hex 🤔, taht would also result in ident safe characters.

Reply:
Does this really matter? The compiler error should make the issue obvious right?

Reply:
I doubt the compiler error would make it obvious, but I'm also not sure if it's a problem. Should be simple to test though to see what the error would be.

Reply:
If needed we should be able to convert the string to an Ident in a way that won't panic by doing: 
```rust
syn::parse_str::<syn::Ident>(...)
```

Reply:
Thinking about this from a different perspective, I guess a downside of this constraint is that a valid meta key is prevented from being used, such as ones with a hyphen, which doesn't make much sense. 🤔

Reply:
How about we convert the key to hex, and use that in the identifier? It'll mean it is always valid, no matter what the key is.

```suggestion
        let ident = format_ident!("__CONTRACT_KEY_{}", key.as_bytes().iter().map(|b| format!("{b:02x}")).collect::<String>())
```

Reply:
Hmm I thought it'd be ideal if the variable name visibly made sense, but I guess there's no reason to actually use the variable. I'll take your suggestion.